File: medium/050_Pow_x_n/pow.py
class Solution(object):
    def pow(self, x, n):
        base = x
        result = 1
        while n > 0:
            if n & 1 == 1:
                result *= base
            n >>= 1
            base *= base
        return result

    def myPow(self, x, n):
        """
        :type x: float
        :type n: int
        :rtype: float
        """
        if n == 0:
            return 1
        if x == 0:
            return 0

        if n < 0:
            negative = True
        else:
            negative = False

        n = abs(n)
        result = self.pow(x, n)
        if negative:
            result = 1/result

        return result

solution = Solution()

x, n = 0, 0
result = solution.myPow(x, n)
print "{}**{}, get {}, should be {}".format(x, n, result, x**n)

x, n = 0, 1
result = solution.myPow(x, n)
print "{}**{}, get {}, should be {}".format(x, n, result, x**n)

x, n = 2, 1
result = solution.myPow(x, n)
print "{}**{}, get {}, should be {}".format(x, n, result, x**n)

x, n = 2, 13
result = solution.myPow(x, n)
print "{}**{}, get {}, should be {}".format(x, n, result, x**n)

File: medium/050_Pow_x_n/pow.go
package main

import "fmt"

func pow(x float64, n int) float64 {
	f := x
	var res float64 = 1
	for {
		if n <= 0 {
			break
		}
		i := n & 0x1
		if i == 1 {
			res *= f
		}
		f = f * f
		n >>= 1
	}
	return res
}

func myPow(x float64, n int) float64 {
	if n < 0 {
		return 1 / pow(x, -n)
	} else {
		return pow(x, n)
	}
}

func myPowBF(x float64, n int) float64 {
	var m int
	if n < 0 {
		m = -n
	} else {
		m = n
	}
	var res float64 = 1
	for i := 0; i < m; i++ {
		res *= x
	}
	if n < 0 {
		return 1 / res
	} else {
		return res
	}
}

func testMyPow(x float64, n int) {
	res := myPow(x, n)
	ans := myPowBF(x, n)
	if res != ans {
		panic(fmt.Sprintf("Fail on %f**%d, get %f, should %f", x, n, res, ans))
	}
}

func main() {
	testMyPow(2, 0)
	testMyPow(2, 1)
	testMyPow(2, 2)
	testMyPow(2, 3)
	testMyPow(2, 4)
	testMyPow(2, 5)
	testMyPow(2, 6)
	testMyPow(2, 7)
	testMyPow(2, 8)
	testMyPow(2, 9)
	testMyPow(2, -1)
	testMyPow(2, -2)
}
